nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py

```diff
@@ -0,0 +1,365 @@
+        cuda.syncthreads()
+
+    # After final sync, alphas[b, T-1, U - 1] + logprobs[b, T-1, U-1, blank] + denom[b, T-1, U-1] gives
+    # log-likelihood of forward pass.
+    if u == 0:
+        loglike = alphas[offset + (T - 1) * maxU + U - 1] + logp(
+            denom, acts, maxT, maxU, alphabet_size, b, T - 1, U - 1, blank_
+        )
+        llForward[b] = loglike
+
+
+@cuda.jit()
+def compute_betas_kernel(
+    acts: torch.Tensor,
+    denom: torch.Tensor,
+    betas: torch.Tensor,
+    llBackward: torch.Tensor,
+    xlen: torch.Tensor,
+    ylen: torch.Tensor,
+    mlabels: torch.Tensor,  # [B, U]
+    minibatch: int,
+    maxT: int,
+    maxU: int,
+    alphabet_size: int,
+    blank_: int,
+):
+    """
+    Compute beta (backward variable) probabilities over the transduction step.
+
+    Args:
+        acts: Tensor of shape [B, T, U, V+1] flattened. Represents the logprobs activation tensor.
+        denom: Tensor of shape [B, T, U] flattened. Represents the denominator of the logprobs activation tensor
+            across entire vocabulary.
+        betas: Zero tensor of shape [B, T, U]. Will be updated inside the kernel with the backward variable
+            probabilities.
+        llBackward: Zero tensor of shape [B]. Represents the log-likelihood of the backward pass.
+            Returned as the backward pass loss that is reduced by the optimizer.
+        xlen: Vector of length B which contains the actual acoustic sequence lengths in the padded
+            activation tensor.
+        ylen: Vector of length B which contains the actual target sequence lengths in the padded
+            activation tensor.
+        mlabels: Matrix of shape [B, U+1] (+1 here is due to <SOS> token - usually the RNNT blank).
+            The matrix contains the padded target transcription that must be predicted.
+        minibatch: Int representing the batch size.
+        maxT: The maximum possible acoustic sequence length. Represents T in the logprobs tensor.
+        maxU: The maximum possible target sequence length. Represents U in the logprobs tensor.
+        alphabet_size: The vocabulary dimension V+1 (inclusive of RNNT blank).
+        blank_: Index of the RNNT blank token in the vocabulary. Generally the first or last token in the vocab.
+
+    Updates:
+        Kernel inplace updates the following inputs:
+        -   betas: backward variable scores.
+        -   llBackward: log-likelihood of backward variable.
+    """
+    # // launch B blocks, each block has U threads
+    b = cuda.blockIdx.x  # // batch id
+    u = cuda.threadIdx.x  # label id, u
+    T = xlen[b]  # select AM length of current sample
+    U = ylen[b] + 1  # select target length of current sample, +1 for the blank token
+
+    labels: torch.Tensor = mlabels[b]  # mb label start point, equivalent to mlabels + b * (maxU - 1)
+    offset = b * maxT * maxU  # pointer indexing offset
+
+    # betas += offset # pointer offset, ignored since we explicitly add offset
```

Comment:
Should we keep it?

Reply:
Yeah, I would like to keep it since otherwise its hard to understand what offset is really doing (pointer indexing)